ext/json/json_encoder.c

```diff
@@ -71,7 +71,14 @@ static inline void php_json_pretty_print_indent(smart_str *buf, int options, php
 
 /* }}} */
 
-static inline bool php_json_is_valid_double(double d) /* {{{ */
+static
+#if defined(_MSC_VER) && defined(_M_ARM64)
+// MSVC bug: https://developercommunity.visualstudio.com/t/corrupt-optimization-on-arm64-with-Ox-/10102551
```

Comment:
Nice to have a comment referring to the bug report here! :)

Reply:
@cmb69 this bug also affects x86_64 at VS 2022 17.2, should we remove `&& defined(_M_ARM64)` ?

see [ruby issue](https://bugs.ruby-lang.org/issues/18362) [ruby workaround](https://github.com/ruby/ruby/commit/ed8776f96034b245b5418d4817d0ebda3db24198) [vs community issue](https://developercommunity.visualstudio.com/t/Prev-Issue---with-__assume-isnan-/1597317)

generated asm code from vs 17.2 on my environment for x86_64:
```
$LN14:
  000000000000027C: F2 0F 10 07        movsd       xmm0,mmword ptr [rdi]
  0000000000000280: E8 00 00 00 00     call        _dclass
  0000000000000285: F2 0F 10 0F        movsd       xmm1,mmword ptr [rdi]
  0000000000000289: 48 8D 0D 00 00 00  lea         rcx,[$SG72158]
                    00
  0000000000000290: 45 33 C0           xor         r8d,r8d
  0000000000000293: 66 48 0F 7E CA     movq        rdx,xmm1
  0000000000000298: 66 83 F8 01        cmp         ax,1 <- only checked infinite
  000000000000029C: 41 0F 95 C0        setne       r8b
  00000000000002A0: E8 00 00 00 00     call        printf
  00000000000002A5: F2 0F 10 07        movsd       xmm0,mmword ptr [rdi]
  00000000000002A9: E8 00 00 00 00     call        _dclass
  00000000000002AE: 66 83 F8 01        cmp         ax,1
  00000000000002B2: 74 16              je          00000000000002CA
  00000000000002B4: F2 0F 10 0F        movsd       xmm1,mmword ptr [rdi]
  00000000000002B8: 44 8B C5           mov         r8d,ebp
  00000000000002BB: 48 8B CB           mov         rcx,rbx
  00000000000002BE: E8 00 00 00 00     call        php_json_encode_double
  00000000000002C3: 33 C0              xor         eax,eax
  00000000000002C5: E9 E7 FD FF FF     jmp         00000000000000B1
  00000000000002CA: C7 46 08 07 00 00  mov         dword ptr [rsi+8],7
                    00
  00000000000002D1: BF 01 00 00 00     mov         edi,1
  00000000000002D6: 48 8B 03           mov         rax,qword ptr [rbx]
  00000000000002D9: 48 85 C0           test        rax,rax
  00000000000002DC: 74 0D              je          00000000000002EB
  00000000000002DE: 48 8B 78 10        mov         rdi,qword ptr [rax+10h]
  00000000000002E2: 48 FF C7           inc         rdi
  00000000000002E5: 48 3B 7B 08        cmp         rdi,qword ptr [rbx+8]
  00000000000002E9: 72 0B              jb          00000000000002F6
  00000000000002EB: 48 8B D7           mov         rdx,rdi
  00000000000002EE: 48 8B CB           mov         rcx,rbx
  00000000000002F1: E8 00 00 00 00     call        smart_str_erealloc@@16
  00000000000002F6: 48 8B 03           mov         rax,qword ptr [rbx]
  00000000000002F9: C6 44 38 17 30     mov         byte ptr [rax+rdi+17h],30h
  00000000000002FE: 48 8B 03           mov         rax,qword ptr [rbx]
  0000000000000301: 48 89 78 10        mov         qword ptr [rax+10h],rdi
  0000000000000305: 33 C0              xor         eax,eax
  0000000000000307: E9 A5 FD FF FF     jmp         00000000000000B1
```

correct code generated by vs 17.3 preview:
```
$LN14:
  000000000000027C: 0F 29 74 24 60     movaps      xmmword ptr [rsp+60h],xmm6
  0000000000000281: F2 0F 10 36        movsd       xmm6,mmword ptr [rsi]
  0000000000000285: 0F 28 C6           movaps      xmm0,xmm6
  0000000000000288: E8 00 00 00 00     call        _dclass
  000000000000028D: BF 01 00 00 00     mov         edi,1
  0000000000000292: 66 3B C7           cmp         ax,di
  0000000000000295: 74 11              je          00000000000002A8
  0000000000000297: 66 0F 2E F6        ucomisd     xmm6,xmm6
  000000000000029B: 0F 9A C0           setp        al
  000000000000029E: 40 3A C7           cmp         al,dil
  00000000000002A1: 74 05              je          00000000000002A8
  00000000000002A3: 44 8B C7           mov         r8d,edi
  00000000000002A6: EB 03              jmp         00000000000002AB
  00000000000002A8: 45 33 C0           xor         r8d,r8d
  00000000000002AB: 0F 28 CE           movaps      xmm1,xmm6
  00000000000002AE: 48 8D 0D 00 00 00  lea         rcx,[$SG72170]
                    00
  00000000000002B5: 66 48 0F 7E F2     movq        rdx,xmm6
  00000000000002BA: E8 00 00 00 00     call        printf
  00000000000002BF: F2 0F 10 36        movsd       xmm6,mmword ptr [rsi]
  00000000000002C3: 0F 28 C6           movaps      xmm0,xmm6
  00000000000002C6: E8 00 00 00 00     call        _dclass
  00000000000002CB: 66 3B C7           cmp         ax,di
  00000000000002CE: 74 26              je          00000000000002F6
  00000000000002D0: 66 0F 2E F6        ucomisd     xmm6,xmm6
  00000000000002D4: 0F 9A C0           setp        al <-- checked NaN also
  00000000000002D7: 40 3A C7           cmp         al,dil
  00000000000002DA: 74 1A              je          00000000000002F6
  00000000000002DC: 45 8B C6           mov         r8d,r14d
  00000000000002DF: 0F 28 CE           movaps      xmm1,xmm6
  00000000000002E2: 48 8B CB           mov         rcx,rbx
  00000000000002E5: E8 00 00 00 00     call        php_json_encode_double
  00000000000002EA: 0F 28 74 24 60     movaps      xmm6,xmmword ptr [rsp+60h]
  00000000000002EF: 33 C0              xor         eax,eax
  00000000000002F1: E9 BC FD FF FF     jmp         00000000000000B2
  00000000000002F6: C7 45 08 07 00 00  mov         dword ptr [rbp+8],7
                    00
  00000000000002FD: 48 8B 03           mov         rax,qword ptr [rbx]
  0000000000000300: 48 85 C0           test        rax,rax
  0000000000000303: 74 0D              je          0000000000000312
  0000000000000305: 48 8B 78 10        mov         rdi,qword ptr [rax+10h]
  0000000000000309: 48 FF C7           inc         rdi
  000000000000030C: 48 3B 7B 08        cmp         rdi,qword ptr [rbx+8]
  0000000000000310: 72 0B              jb          000000000000031D
  0000000000000312: 48 8B D7           mov         rdx,rdi
  0000000000000315: 48 8B CB           mov         rcx,rbx
  0000000000000318: E8 00 00 00 00     call        smart_str_erealloc@@16
  000000000000031D: 48 8B 03           mov         rax,qword ptr [rbx]
  0000000000000320: 0F 28 74 24 60     movaps      xmm6,xmmword ptr [rsp+60h]
  0000000000000325: C6 44 38 17 30     mov         byte ptr [rax+rdi+17h],30h
  000000000000032A: 48 8B 03           mov         rax,qword ptr [rbx]
  000000000000032D: 48 89 78 10        mov         qword ptr [rax+10h],rdi
  0000000000000331: 33 C0              xor         eax,eax
  0000000000000333: E9 7A FD FF FF     jmp         00000000000000B2
```

Reply:
Well, given that the official builds still use VS 2019, and that issue is fixed with 17.3, I think we can leave that as is.